helix-term/src/ui/document.rs

```diff
@@ -0,0 +1,207 @@
+        text_render: &mut TextRender,
+    ) -> Self {
+        let mut render = DocumentRender {
+            config,
+            theme,
+            text,
+            highlights,
+            cursor: DocumentCursor::new(text, offset.row, char_offset, config, &viewport),
+            // render: TextRender::new(surface, render_config, offset.col, viewport),
+            spans: Vec::with_capacity(64),
+            finished: false,
+        };
+
+        // advance to first highlight scope
+        render.advance_highlight_scope(text_render);
+        render
+    }
+
+    /// Returns the line in the doucment that will be rendered next
+    pub fn doc_line(&self) -> usize {
+        self.cursor.doc_position().row
+    }
+
+    fn advance_highlight_scope(&mut self, text_render: &mut TextRender) {
+        while let Some(event) = self.highlights.next() {
+            match event {
+                HighlightEvent::HighlightStart(span) => self.spans.push(span),
+                HighlightEvent::HighlightEnd => {
+                    self.spans.pop();
+                }
+                HighlightEvent::Source { start, end } => {
+                    if start == end {
+                        continue;
+                    }
+                    // TODO cursor end
+                    let style = self
+                        .spans
+                        .iter()
+                        .fold(text_render.config.text_style, |acc, span| {
+                            acc.patch(self.theme.highlight(span.0))
+                        });
+                    self.cursor.set_highlight_scope(start, end, style);
+                    return;
+                }
+            }
+        }
+        self.finished = true;
+    }
+
+    /// Perform a softwrap
+    fn wrap_line(&mut self, text_render: &mut TextRender) {
+        // Fully wrapping this word would wrap too much wrap
+        // wrap inside the word instead
+        if self.cursor.word_width() > self.cursor.config.max_wrap {
+            self.cursor.take_word_buf_until(|grapheme| {
+                if text_render.space_left() < grapheme.min_width() as usize {
+                    // leave this grapheme in the word_buf
+                    Some(grapheme)
+                } else {
+                    text_render.draw_grapheme(grapheme);
+                    None
+                }
+            });
+        }
+
+        let indent_level = match text_render.indent_level() {
+            IndentLevel::Known(indent_level)
+                if indent_level <= self.cursor.config.max_indent_retain =>
+            {
+                IndentLevel::Known(indent_level)
+            }
+            _ => IndentLevel::None,
+        };
+        self.advance_line(indent_level, text_render);
+        text_render.skip(self.config.soft_wrap.wrap_indent);
+    }
+
+    /// Returns whether this document renderer finished rendering
+    /// either because the viewport end or EOF was reached
+    pub fn finished(&self) -> bool {
```

Comment:
Nit: let's use `is_finished`, stdlib usually uses the `is_` prefix for these types of methods